// @digdir/designsystemet/dist/src/scripts/update-template.js
// src/scripts/update-template.ts
import path from "path";
import pc2 from "picocolors";
import * as R from "ramda";

// ../../internal/design-tokens/semantic/color.json
var color_default = {
  color: {
    accent: {
      "background-default": {
        $type: "color",
        $value: "{color.accent.1}"
      },
      "background-tinted": {
        $type: "color",
        $value: "{color.accent.2}"
      },
      "surface-default": {
        $type: "color",
        $value: "{color.accent.3}"
      },
      "surface-tinted": {
        $type: "color",
        $value: "{color.accent.4}"
      },
      "surface-hover": {
        $type: "color",
        $value: "{color.accent.5}"
      },
      "surface-active": {
        $type: "color",
        $value: "{color.accent.6}"
      },
      "border-subtle": {
        $type: "color",
        $value: "{color.accent.7}"
      },
      "border-default": {
        $type: "color",
        $value: "{color.accent.8}"
      },
      "border-strong": {
        $type: "color",
        $value: "{color.accent.9}"
      },
      "text-subtle": {
        $type: "color",
        $value: "{color.accent.10}"
      },
      "text-default": {
        $type: "color",
        $value: "{color.accent.11}"
      },
      "base-default": {
        $type: "color",
        $value: "{color.accent.12}"
      },
      "base-hover": {
        $type: "color",
        $value: "{color.accent.13}"
      },
      "base-active": {
        $type: "color",
        $value: "{color.accent.14}"
      },
      "base-contrast-subtle": {
        $type: "color",
        $value: "{color.accent.15}"
      },
      "base-contrast-default": {
        $type: "color",
        $value: "{color.accent.16}"
      }
    },
    neutral: {
      "background-default": {
        $type: "color",
        $value: "{color.neutral.1}"
      },
      "background-tinted": {
        $type: "color",
        $value: "{color.neutral.2}"
      },
      "surface-default": {
        $type: "color",
        $value: "{color.neutral.3}"
      },
      "surface-tinted": {
        $type: "color",
        $value: "{color.neutral.4}"
      },
      "surface-hover": {
        $type: "color",
        $value: "{color.neutral.5}"
      },
      "surface-active": {
        $type: "color",
        $value: "{color.neutral.6}"
      },
      "border-subtle": {
        $type: "color",
        $value: "{color.neutral.7}"
      },
      "border-default": {
        $type: "color",
        $value: "{color.neutral.8}"
      },
      "border-strong": {
        $type: "color",
        $value: "{color.neutral.9}"
      },
      "text-subtle": {
        $type: "color",
        $value: "{color.neutral.10}"
      },
      "text-default": {
        $type: "color",
        $value: "{color.neutral.11}"
      },
      "base-default": {
        $type: "color",
        $value: "{color.neutral.12}"
      },
      "base-hover": {
        $type: "color",
        $value: "{color.neutral.13}"
      },
      "base-active": {
        $type: "color",
        $value: "{color.neutral.14}"
      },
      "base-contrast-subtle": {
        $type: "color",
        $value: "{color.neutral.15}"
      },
      "base-contrast-default": {
        $type: "color",
        $value: "{color.neutral.16}"
      }
    },
    brand1: {
      "background-default": {
        $type: "color",
        $value: "{color.brand1.1}"
      },
      "background-tinted": {
        $type: "color",
        $value: "{color.brand1.2}"
      },
      "surface-default": {
        $type: "color",
        $value: "{color.brand1.3}"
      },
      "surface-tinted": {
        $type: "color",
        $value: "{color.brand1.4}"
      },
      "surface-hover": {
        $type: "color",
        $value: "{color.brand1.5}"
      },
      "surface-active": {
        $type: "color",
        $value: "{color.brand1.6}"
      },
      "border-subtle": {
        $type: "color",
        $value: "{color.brand1.7}"
      },
      "border-default": {
        $type: "color",
        $value: "{color.brand1.8}"
      },
      "border-strong": {
        $type: "color",
        $value: "{color.brand1.9}"
      },
      "text-subtle": {
        $type: "color",
        $value: "{color.brand1.10}"
      },
      "text-default": {
        $type: "color",
        $value: "{color.brand1.11}"
      },
      "base-default": {
        $type: "color",
        $value: "{color.brand1.12}"
      },
      "base-hover": {
        $type: "color",
        $value: "{color.brand1.13}"
      },
      "base-active": {
        $type: "color",
        $value: "{color.brand1.14}"
      },
      "base-contrast-subtle": {
        $type: "color",
        $value: "{color.brand1.15}"
      },
      "base-contrast-default": {
        $type: "color",
        $value: "{color.brand1.16}"
      }
    },
    brand2: {
      "background-default": {
        $type: "color",
        $value: "{color.brand2.1}"
      },
      "background-tinted": {
        $type: "color",
        $value: "{color.brand2.2}"
      },
      "surface-default": {
        $type: "color",
        $value: "{color.brand2.3}"
      },
      "surface-tinted": {
        $type: "color",
        $value: "{color.brand2.4}"
      },
      "surface-hover": {
        $type: "color",
        $value: "{color.brand2.5}"
      },
      "surface-active": {
        $type: "color",
        $value: "{color.brand2.6}"
      },
      "border-subtle": {
        $type: "color",
        $value: "{color.brand2.7}"
      },
      "border-default": {
        $type: "color",
        $value: "{color.brand2.8}"
      },
      "border-strong": {
        $type: "color",
        $value: "{color.brand2.9}"
      },
      "text-subtle": {
        $type: "color",
        $value: "{color.brand2.10}"
      },
      "text-default": {
        $type: "color",
        $value: "{color.brand2.11}"
      },
      "base-default": {
        $type: "color",
        $value: "{color.brand2.12}"
      },
      "base-hover": {
        $type: "color",
        $value: "{color.brand2.13}"
      },
      "base-active": {
        $type: "color",
        $value: "{color.brand2.14}"
      },
      "base-contrast-subtle": {
        $type: "color",
        $value: "{color.brand2.15}"
      },
      "base-contrast-default": {
        $type: "color",
        $value: "{color.brand2.16}"
      }
    },
    brand3: {
      "background-default": {
        $type: "color",
        $value: "{color.brand3.1}"
      },
      "background-tinted": {
        $type: "color",
        $value: "{color.brand3.2}"
      },
      "surface-default": {
        $type: "color",
        $value: "{color.brand3.3}"
      },
      "surface-tinted": {
        $type: "color",
        $value: "{color.brand3.4}"
      },
      "surface-hover": {
        $type: "color",
        $value: "{color.brand3.5}"
      },
      "surface-active": {
        $type: "color",
        $value: "{color.brand3.6}"
      },
      "border-subtle": {
        $type: "color",
        $value: "{color.brand3.7}"
      },
      "border-default": {
        $type: "color",
        $value: "{color.brand3.8}"
      },
      "border-strong": {
        $type: "color",
        $value: "{color.brand3.9}"
      },
      "text-subtle": {
        $type: "color",
        $value: "{color.brand3.10}"
      },
      "text-default": {
        $type: "color",
        $value: "{color.brand3.11}"
      },
      "base-default": {
        $type: "color",
        $value: "{color.brand3.12}"
      },
      "base-hover": {
        $type: "color",
        $value: "{color.brand3.13}"
      },
      "base-active": {
        $type: "color",
        $value: "{color.brand3.14}"
      },
      "base-contrast-subtle": {
        $type: "color",
        $value: "{color.brand3.15}"
      },
      "base-contrast-default": {
        $type: "color",
        $value: "{color.brand3.16}"
      }
    },
    info: {
      "background-default": {
        $type: "color",
        $value: "{color.info.1}"
      },
      "background-tinted": {
        $type: "color",
        $value: "{color.info.2}"
      },
      "surface-default": {
        $type: "color",
        $value: "{color.info.3}"
      },
      "surface-tinted": {
        $type: "color",
        $value: "{color.info.4}"
      },
      "surface-hover": {
        $type: "color",
        $value: "{color.info.5}"
      },
      "surface-active": {
        $type: "color",
        $value: "{color.info.6}"
      },
      "border-subtle": {
        $type: "color",
        $value: "{color.info.7}"
      },
      "border-default": {
        $type: "color",
        $value: "{color.info.8}"
      },
      "border-strong": {
        $type: "color",
        $value: "{color.info.9}"
      },
      "text-subtle": {
        $type: "color",
        $value: "{color.info.10}"
      },
      "text-default": {
        $type: "color",
        $value: "{color.info.11}"
      },
      "base-default": {
        $type: "color",
        $value: "{color.info.12}"
      },
      "base-hover": {
        $type: "color",
        $value: "{color.info.13}"
      },
      "base-active": {
        $type: "color",
        $value: "{color.info.14}"
      },
      "base-contrast-subtle": {
        $type: "color",
        $value: "{color.info.15}"
      },
      "base-contrast-default": {
        $type: "color",
        $value: "{color.info.16}"
      }
    },
    success: {
      "background-default": {
        $type: "color",
        $value: "{color.success.1}"
      },
      "background-tinted": {
        $type: "color",
        $value: "{color.success.2}"
      },
      "surface-default": {
        $type: "color",
        $value: "{color.success.3}"
      },
      "surface-tinted": {
        $type: "color",
        $value: "{color.success.4}"
      },
      "surface-hover": {
        $type: "color",
        $value: "{color.success.5}"
      },
      "surface-active": {
        $type: "color",
        $value: "{color.success.6}"
      },
      "border-subtle": {
        $type: "color",
        $value: "{color.success.7}"
      },
      "border-default": {
        $type: "color",
        $value: "{color.success.8}"
      },
      "border-strong": {
        $type: "color",
        $value: "{color.success.9}"
      },
      "text-subtle": {
        $type: "color",
        $value: "{color.success.10}"
      },
      "text-default": {
        $type: "color",
        $value: "{color.success.11}"
      },
      "base-default": {
        $type: "color",
        $value: "{color.success.12}"
      },
      "base-hover": {
        $type: "color",
        $value: "{color.success.13}"
      },
      "base-active": {
        $type: "color",
        $value: "{color.success.14}"
      },
      "base-contrast-subtle": {
        $type: "color",
        $value: "{color.success.15}"
      },
      "base-contrast-default": {
        $type: "color",
        $value: "{color.success.16}"
      }
    },
    warning: {
      "background-default": {
        $type: "color",
        $value: "{color.warning.1}"
      },
      "background-tinted": {
        $type: "color",
        $value: "{color.warning.2}"
      },
      "surface-default": {
        $type: "color",
        $value: "{color.warning.3}"
      },
      "surface-tinted": {
        $type: "color",
        $value: "{color.warning.4}"
      },
      "surface-hover": {
        $type: "color",
        $value: "{color.warning.5}"
      },
      "surface-active": {
        $type: "color",
        $value: "{color.warning.6}"
      },
      "border-subtle": {
        $type: "color",
        $value: "{color.warning.7}"
      },
      "border-default": {
        $type: "color",
        $value: "{color.warning.8}"
      },
      "border-strong": {
        $type: "color",
        $value: "{color.warning.9}"
      },
      "text-subtle": {
        $type: "color",
        $value: "{color.warning.10}"
      },
      "text-default": {
        $type: "color",
        $value: "{color.warning.11}"
      },
      "base-default": {
        $type: "color",
        $value: "{color.warning.12}"
      },
      "base-hover": {
        $type: "color",
        $value: "{color.warning.13}"
      },
      "base-active": {
        $type: "color",
        $value: "{color.warning.14}"
      },
      "base-contrast-subtle": {
        $type: "color",
        $value: "{color.warning.15}"
      },
      "base-contrast-default": {
        $type: "color",
        $value: "{color.warning.16}"
      }
    },
    danger: {
      "background-default": {
        $type: "color",
        $value: "{color.danger.1}"
      },
      "background-tinted": {
        $type: "color",
        $value: "{color.danger.2}"
      },
      "surface-default": {
        $type: "color",
        $value: "{color.danger.3}"
      },
      "surface-tinted": {
        $type: "color",
        $value: "{color.danger.4}"
      },
      "surface-hover": {
        $type: "color",
        $value: "{color.danger.5}"
      },
      "surface-active": {
        $type: "color",
        $value: "{color.danger.6}"
      },
      "border-subtle": {
        $type: "color",
        $value: "{color.danger.7}"
      },
      "border-default": {
        $type: "color",
        $value: "{color.danger.8}"
      },
      "border-strong": {
        $type: "color",
        $value: "{color.danger.9}"
      },
      "text-subtle": {
        $type: "color",
        $value: "{color.danger.10}"
      },
      "text-default": {
        $type: "color",
        $value: "{color.danger.11}"
      },
      "base-default": {
        $type: "color",
        $value: "{color.danger.12}"
      },
      "base-hover": {
        $type: "color",
        $value: "{color.danger.13}"
      },
      "base-active": {
        $type: "color",
        $value: "{color.danger.14}"
      },
      "base-contrast-subtle": {
        $type: "color",
        $value: "{color.danger.15}"
      },
      "base-contrast-default": {
        $type: "color",
        $value: "{color.danger.16}"
      }
    },
    focus: {
      inner: {
        $type: "color",
        $value: "{color.neutral.background-default}"
      },
      outer: {
        $type: "color",
        $value: "{color.neutral.text-default}"
      }
    }
  },
  link: {
    color: {
      visited: {
        $type: "color",
        $value: "{color.link.visited}"
      }
    }
  }
};

// ../../internal/design-tokens/semantic/modes/main-color/accent.json
var accent_default = {
  color: {
    main: {
      "background-default": {
        $type: "color",
        $value: "{color.accent.background-default}"
      },
      "background-tinted": {
        $type: "color",
        $value: "{color.accent.background-tinted}"
      },
      "surface-default": {
        $type: "color",
        $value: "{color.accent.surface-default}"
      },
      "surface-tinted": {
        $type: "color",
        $value: "{color.accent.surface-tinted}"
      },
      "surface-hover": {
        $type: "color",
        $value: "{color.accent.surface-hover}"
      },
      "surface-active": {
        $type: "color",
        $value: "{color.accent.surface-active}"
      },
      "border-subtle": {
        $type: "color",
        $value: "{color.accent.border-subtle}"
      },
      "border-default": {
        $type: "color",
        $value: "{color.accent.border-default}"
      },
      "border-strong": {
        $type: "color",
        $value: "{color.accent.border-strong}"
      },
      "text-subtle": {
        $type: "color",
        $value: "{color.accent.text-subtle}"
      },
      "text-default": {
        $type: "color",
        $value: "{color.accent.text-default}"
      },
      "base-default": {
        $type: "color",
        $value: "{color.accent.base-default}"
      },
      "base-hover": {
        $type: "color",
        $value: "{color.accent.base-hover}"
      },
      "base-active": {
        $type: "color",
        $value: "{color.accent.base-active}"
      },
      "base-contrast-subtle": {
        $type: "color",
        $value: "{color.accent.base-contrast-subtle}"
      },
      "base-contrast-default": {
        $type: "color",
        $value: "{color.accent.base-contrast-default}"
      }
    }
  }
};

// ../../internal/design-tokens/themes/digdir.json
var digdir_default = {
  color: {
    accent: {
      "1": {
        $type: "color",
        $value: "{digdir.accent.1}"
      },
      "2": {
        $type: "color",
        $value: "{digdir.accent.2}"
      },
      "3": {
        $type: "color",
        $value: "{digdir.accent.3}"
      },
      "4": {
        $type: "color",
        $value: "{digdir.accent.4}"
      },
      "5": {
        $type: "color",
        $value: "{digdir.accent.5}"
      },
      "6": {
        $type: "color",
        $value: "{digdir.accent.6}"
      },
      "7": {
        $type: "color",
        $value: "{digdir.accent.7}"
      },
      "8": {
        $type: "color",
        $value: "{digdir.accent.8}"
      },
      "9": {
        $type: "color",
        $value: "{digdir.accent.9}"
      },
      "10": {
        $type: "color",
        $value: "{digdir.accent.10}"
      },
      "11": {
        $type: "color",
        $value: "{digdir.accent.11}"
      },
      "12": {
        $type: "color",
        $value: "{digdir.accent.12}"
      },
      "13": {
        $type: "color",
        $value: "{digdir.accent.13}"
      },
      "14": {
        $type: "color",
        $value: "{digdir.accent.14}"
      },
      "15": {
        $type: "color",
        $value: "{digdir.accent.15}"
      },
      "16": {
        $type: "color",
        $value: "{digdir.accent.16}"
      }
    },
    neutral: {
      "1": {
        $type: "color",
        $value: "{digdir.neutral.1}"
      },
      "2": {
        $type: "color",
        $value: "{digdir.neutral.2}"
      },
      "3": {
        $type: "color",
        $value: "{digdir.neutral.3}"
      },
      "4": {
        $type: "color",
        $value: "{digdir.neutral.4}"
      },
      "5": {
        $type: "color",
        $value: "{digdir.neutral.5}"
      },
      "6": {
        $type: "color",
        $value: "{digdir.neutral.6}"
      },
      "7": {
        $type: "color",
        $value: "{digdir.neutral.7}"
      },
      "8": {
        $type: "color",
        $value: "{digdir.neutral.8}"
      },
      "9": {
        $type: "color",
        $value: "{digdir.neutral.9}"
      },
      "10": {
        $type: "color",
        $value: "{digdir.neutral.10}"
      },
      "11": {
        $type: "color",
        $value: "{digdir.neutral.11}"
      },
      "12": {
        $type: "color",
        $value: "{digdir.neutral.12}"
      },
      "13": {
        $type: "color",
        $value: "{digdir.neutral.13}"
      },
      "14": {
        $type: "color",
        $value: "{digdir.neutral.14}"
      },
      "15": {
        $type: "color",
        $value: "{digdir.neutral.15}"
      },
      "16": {
        $type: "color",
        $value: "{digdir.neutral.16}"
      }
    },
    brand1: {
      "1": {
        $type: "color",
        $value: "{digdir.brand1.1}"
      },
      "2": {
        $type: "color",
        $value: "{digdir.brand1.2}"
      },
      "3": {
        $type: "color",
        $value: "{digdir.brand1.3}"
      },
      "4": {
        $type: "color",
        $value: "{digdir.brand1.4}"
      },
      "5": {
        $type: "color",
        $value: "{digdir.brand1.5}"
      },
      "6": {
        $type: "color",
        $value: "{digdir.brand1.6}"
      },
      "7": {
        $type: "color",
        $value: "{digdir.brand1.7}"
      },
      "8": {
        $type: "color",
        $value: "{digdir.brand1.8}"
      },
      "9": {
        $type: "color",
        $value: "{digdir.brand1.9}"
      },
      "10": {
        $type: "color",
        $value: "{digdir.brand1.10}"
      },
      "11": {
        $type: "color",
        $value: "{digdir.brand1.11}"
      },
      "12": {
        $type: "color",
        $value: "{digdir.brand1.12}"
      },
      "13": {
        $type: "color",
        $value: "{digdir.brand1.13}"
      },
      "14": {
        $type: "color",
        $value: "{digdir.brand1.14}"
      },
      "15": {
        $type: "color",
        $value: "{digdir.brand1.15}"
      },
      "16": {
        $type: "color",
        $value: "{digdir.brand1.16}"
      }
    },
    brand2: {
      "1": {
        $type: "color",
        $value: "{digdir.brand2.1}"
      },
      "2": {
        $type: "color",
        $value: "{digdir.brand2.2}"
      },
      "3": {
        $type: "color",
        $value: "{digdir.brand2.3}"
      },
      "4": {
        $type: "color",
        $value: "{digdir.brand2.4}"
      },
      "5": {
        $type: "color",
        $value: "{digdir.brand2.5}"
      },
      "6": {
        $type: "color",
        $value: "{digdir.brand2.6}"
      },
      "7": {
        $type: "color",
        $value: "{digdir.brand2.7}"
      },
      "8": {
        $type: "color",
        $value: "{digdir.brand2.8}"
      },
      "9": {
        $type: "color",
        $value: "{digdir.brand2.9}"
      },
      "10": {
        $type: "color",
        $value: "{digdir.brand2.10}"
      },
      "11": {
        $type: "color",
        $value: "{digdir.brand2.11}"
      },
      "12": {
        $type: "color",
        $value: "{digdir.brand2.12}"
      },
      "13": {
        $type: "color",
        $value: "{digdir.brand2.13}"
      },
      "14": {
        $type: "color",
        $value: "{digdir.brand2.14}"
      },
      "15": {
        $type: "color",
        $value: "{digdir.brand2.15}"
      },
      "16": {
        $type: "color",
        $value: "{digdir.brand2.16}"
      }
    },
    brand3: {
      "1": {
        $type: "color",
        $value: "{digdir.brand3.1}"
      },
      "2": {
        $type: "color",
        $value: "{digdir.brand3.2}"
      },
      "3": {
        $type: "color",
        $value: "{digdir.brand3.3}"
      },
      "4": {
        $type: "color",
        $value: "{digdir.brand3.4}"
      },
      "5": {
        $type: "color",
        $value: "{digdir.brand3.5}"
      },
      "6": {
        $type: "color",
        $value: "{digdir.brand3.6}"
      },
      "7": {
        $type: "color",
        $value: "{digdir.brand3.7}"
      },
      "8": {
        $type: "color",
        $value: "{digdir.brand3.8}"
      },
      "9": {
        $type: "color",
        $value: "{digdir.brand3.9}"
      },
      "10": {
        $type: "color",
        $value: "{digdir.brand3.10}"
      },
      "11": {
        $type: "color",
        $value: "{digdir.brand3.11}"
      },
      "12": {
        $type: "color",
        $value: "{digdir.brand3.12}"
      },
      "13": {
        $type: "color",
        $value: "{digdir.brand3.13}"
      },
      "14": {
        $type: "color",
        $value: "{digdir.brand3.14}"
      },
      "15": {
        $type: "color",
        $value: "{digdir.brand3.15}"
      },
      "16": {
        $type: "color",
        $value: "{digdir.brand3.16}"
      }
    },
    info: {
      "1": {
        $type: "color",
        $value: "{digdir.info.1}"
      },
      "2": {
        $type: "color",
        $value: "{digdir.info.2}"
      },
      "3": {
        $type: "color",
        $value: "{digdir.info.3}"
      },
      "4": {
        $type: "color",
        $value: "{digdir.info.4}"
      },
      "5": {
        $type: "color",
        $value: "{digdir.info.5}"
      },
      "6": {
        $type: "color",
        $value: "{digdir.info.6}"
      },
      "7": {
        $type: "color",
        $value: "{digdir.info.7}"
      },
      "8": {
        $type: "color",
        $value: "{digdir.info.8}"
      },
      "9": {
        $type: "color",
        $value: "{digdir.info.9}"
      },
      "10": {
        $type: "color",
        $value: "{digdir.info.10}"
      },
      "11": {
        $type: "color",
        $value: "{digdir.info.11}"
      },
      "12": {
        $type: "color",
        $value: "{digdir.info.12}"
      },
      "13": {
        $type: "color",
        $value: "{digdir.info.13}"
      },
      "14": {
        $type: "color",
        $value: "{digdir.info.14}"
      },
      "15": {
        $type: "color",
        $value: "{digdir.info.15}"
      },
      "16": {
        $type: "color",
        $value: "{digdir.info.16}"
      }
    },
    success: {
      "1": {
        $type: "color",
        $value: "{digdir.success.1}"
      },
      "2": {
        $type: "color",
        $value: "{digdir.success.2}"
      },
      "3": {
        $type: "color",
        $value: "{digdir.success.3}"
      },
      "4": {
        $type: "color",
        $value: "{digdir.success.4}"
      },
      "5": {
        $type: "color",
        $value: "{digdir.success.5}"
      },
      "6": {
        $type: "color",
        $value: "{digdir.success.6}"
      },
      "7": {
        $type: "color",
        $value: "{digdir.success.7}"
      },
      "8": {
        $type: "color",
        $value: "{digdir.success.8}"
      },
      "9": {
        $type: "color",
        $value: "{digdir.success.9}"
      },
      "10": {
        $type: "color",
        $value: "{digdir.success.10}"
      },
      "11": {
        $type: "color",
        $value: "{digdir.success.11}"
      },
      "12": {
        $type: "color",
        $value: "{digdir.success.12}"
      },
      "13": {
        $type: "color",
        $value: "{digdir.success.13}"
      },
      "14": {
        $type: "color",
        $value: "{digdir.success.14}"
      },
      "15": {
        $type: "color",
        $value: "{digdir.success.15}"
      },
      "16": {
        $type: "color",
        $value: "{digdir.success.16}"
      }
    },
    warning: {
      "1": {
        $type: "color",
        $value: "{digdir.warning.1}"
      },
      "2": {
        $type: "color",
        $value: "{digdir.warning.2}"
      },
      "3": {
        $type: "color",
        $value: "{digdir.warning.3}"
      },
      "4": {
        $type: "color",
        $value: "{digdir.warning.4}"
      },
      "5": {
        $type: "color",
        $value: "{digdir.warning.5}"
      },
      "6": {
        $type: "color",
        $value: "{digdir.warning.6}"
      },
      "7": {
        $type: "color",
        $value: "{digdir.warning.7}"
      },
      "8": {
        $type: "color",
        $value: "{digdir.warning.8}"
      },
      "9": {
        $type: "color",
        $value: "{digdir.warning.9}"
      },
      "10": {
        $type: "color",
        $value: "{digdir.warning.10}"
      },
      "11": {
        $type: "color",
        $value: "{digdir.warning.11}"
      },
      "12": {
        $type: "color",
        $value: "{digdir.warning.12}"
      },
      "13": {
        $type: "color",
        $value: "{digdir.warning.13}"
      },
      "14": {
        $type: "color",
        $value: "{digdir.warning.14}"
      },
      "15": {
        $type: "color",
        $value: "{digdir.warning.15}"
      },
      "16": {
        $type: "color",
        $value: "{digdir.warning.16}"
      }
    },
    danger: {
      "1": {
        $type: "color",
        $value: "{digdir.danger.1}"
      },
      "2": {
        $type: "color",
        $value: "{digdir.danger.2}"
      },
      "3": {
        $type: "color",
        $value: "{digdir.danger.3}"
      },
      "4": {
        $type: "color",
        $value: "{digdir.danger.4}"
      },
      "5": {
        $type: "color",
        $value: "{digdir.danger.5}"
      },
      "6": {
        $type: "color",
        $value: "{digdir.danger.6}"
      },
      "7": {
        $type: "color",
        $value: "{digdir.danger.7}"
      },
      "8": {
        $type: "color",
        $value: "{digdir.danger.8}"
      },
      "9": {
        $type: "color",
        $value: "{digdir.danger.9}"
      },
      "10": {
        $type: "color",
        $value: "{digdir.danger.10}"
      },
      "11": {
        $type: "color",
        $value: "{digdir.danger.11}"
      },
      "12": {
        $type: "color",
        $value: "{digdir.danger.12}"
      },
      "13": {
        $type: "color",
        $value: "{digdir.danger.13}"
      },
      "14": {
        $type: "color",
        $value: "{digdir.danger.14}"
      },
      "15": {
        $type: "color",
        $value: "{digdir.danger.15}"
      },
      "16": {
        $type: "color",
        $value: "{digdir.danger.16}"
      }
    },
    link: {
      visited: {
        $type: "color",
        $value: "{digdir.link.visited}"
      }
    }
  },
  "font-family": {
    $type: "fontFamilies",
    $value: "{digdir.font-family}"
  },
  "font-weight": {
    medium: {
      $type: "fontWeights",
      $value: "{digdir.font-weight.medium}"
    },
    semibold: {
      $type: "fontWeights",
      $value: "{digdir.font-weight.semibold}"
    },
    regular: {
      $type: "fontWeights",
      $value: "{digdir.font-weight.regular}"
    }
  },
  "border-radius": {
    "1": {
      $type: "dimension",
      $value: "min({border-radius.base}*0.5,{border-radius.scale})"
    },
    "2": {
      $type: "dimension",
      $value: "min({border-radius.base},{border-radius.scale}*2)"
    },
    "3": {
      $type: "dimension",
      $value: "min({border-radius.base}*2,{border-radius.scale}*5)"
    },
    "4": {
      $type: "dimension",
      $value: "min({border-radius.base}*3,{border-radius.scale}*7)"
    },
    "5": {
      $type: "dimension",
      $value: "{border-radius.base}"
    },
    "6": {
      $type: "dimension",
      $value: "9999"
    },
    base: {
      $type: "dimension",
      $value: "4"
    },
    scale: {
      $type: "dimension",
      $value: "4"
    }
  }
};

// src/utils.ts
import fs from "fs/promises";
import pc from "picocolors";
var mkdir = async (dir, dry) => {
  if (dry) {
    console.log(`${pc.blue("mkdir")} ${dir}`);
    return Promise.resolve();
  }
  const exists = await fs.access(dir, fs.constants.F_OK).then(() => true).catch(() => false);
  if (exists) {
    return Promise.resolve();
  }
  return fs.mkdir(dir, { recursive: true });
};
var writeFile = async (path2, data, dry) => {
  if (dry) {
    console.log(`${pc.blue("writeFile")} ${path2}`);
    return Promise.resolve();
  }
  return fs.writeFile(path2, data, { encoding: "utf-8" }).catch((error) => {
    console.error(pc.red(`Error writing file: ${path2}`));
    console.error(pc.red(error));
    throw error;
  });
};
var cp = async (src, dest, dry, filter) => {
  if (dry) {
    console.log(`${pc.blue("cp")} ${src} ${dest}`);
    return Promise.resolve();
  }
  return fs.cp(src, dest, { recursive: true, filter });
};
var cleanDir = async (dir, dry) => {
  if (dry) {
    console.log(`${pc.blue("cleanDir")} ${dir}`);
    return Promise.resolve();
  }
  console.log(`
\u{1F525} Cleaning dir ${pc.red(`${dir.trim()}`)} `);
  return fs.rm(dir, { recursive: true, force: true });
};
var readFile = async (path2, dry, allowFileNotFound) => {
  if (dry) {
    console.log(`${pc.blue("readFile")} ${path2}`);
    return Promise.resolve("");
  }
  try {
    return await fs.readFile(path2, "utf-8");
  } catch (error) {
    if (allowFileNotFound && error.code === "ENOENT") {
      return "";
    }
    throw error;
  }
};

// src/scripts/update-template.ts
var DIRNAME = import.meta.dirname || __dirname;
var SOURCE_FILES_PATH = path.join(DIRNAME, "../../../../internal/design-tokens");
var TEMPLATE_FILES_PATH = path.join(DIRNAME, "../tokens/template/design-tokens");
var designTokensPath = (_path) => path.join(SOURCE_FILES_PATH, _path);
var argsFromToPaths = (path_) => [
  path.join(SOURCE_FILES_PATH, path_),
  path.join(TEMPLATE_FILES_PATH, path_)
];
var updateTemplates = async () => {
  await cleanDir(TEMPLATE_FILES_PATH);
  await cp(...argsFromToPaths("primitives/globals.json"));
  await cp(...argsFromToPaths("primitives/modes/size"));
  await cp(...argsFromToPaths("primitives/modes/typography/size"));
  await cp(...argsFromToPaths("semantic/style.json"));
  const digdirTypographyJson = await readFile(designTokensPath("primitives/modes/typography/primary/digdir.json"));
  const typographyTemplate = R.set(
    R.lensPath(["digdir", "font-family", "$value"]),
    "<font-family>",
    JSON.parse(digdirTypographyJson)
  );
  const typographyDir = path.join(TEMPLATE_FILES_PATH, "primitives/modes/typography/");
  await mkdir(typographyDir);
  await writeFile(
    path.join(typographyDir, "typography.template.json"),
    JSON.stringify(typographyTemplate, null, 2).replaceAll("digdir", "<theme>")
  );
  const categoryColorTemplate = accent_default.color.main;
  const categoryDir = path.join(TEMPLATE_FILES_PATH, "semantic/modes");
  await mkdir(categoryDir);
  await writeFile(
    path.join(categoryDir, "color.template.json"),
    JSON.stringify(categoryColorTemplate, null, 2).replaceAll("color.accent", "color.<color>")
  );
  const colorBaseFile = {
    ...color_default,
    // Remove custom colors and severity colors as they are defined by theme
    color: R.pick(["focus"], color_default.color),
    link: {
      color: {
        visited: {
          $type: "color",
          $value: "{color.link.visited}"
        }
      }
    }
  };
  await writeFile(
    path.join(TEMPLATE_FILES_PATH, `semantic/color.base.template.json`),
    JSON.stringify(colorBaseFile, null, 2)
  );
  const semanticColorTemplate = color_default.color.accent;
  await writeFile(
    path.join(TEMPLATE_FILES_PATH, `semantic/color.template.json`),
    JSON.stringify(semanticColorTemplate, null, 2).replaceAll("color.accent", "color.<color>")
  );
  const themeBaseFile = {
    color: {},
    ...R.omit(["color"], digdir_default)
  };
  await mkdir(path.join(TEMPLATE_FILES_PATH, "themes"));
  await writeFile(
    path.join(TEMPLATE_FILES_PATH, `themes/theme.base.template.json`),
    JSON.stringify(themeBaseFile, null, 2).replaceAll("digdir", "<theme>")
  );
  const themeColorTemplate = digdir_default.color.accent;
  await writeFile(
    path.join(TEMPLATE_FILES_PATH, `themes/theme.template.json`),
    JSON.stringify(themeColorTemplate, null, 2).replaceAll("digdir.accent", "<theme>.<color>")
  );
  console.log(pc2.green("Templates updated"));
};
await updateTemplates();
export {
  updateTemplates
};
