// @digdir/designsystemet/dist/src/tokens/create/defaults.js
// src/tokens/create/defaults.ts
import * as R from "ramda";

// src/tokens/template/design-tokens/primitives/globals.json with { type: 'json' }
var globals_default = {
  "border-width": {
    "1": {
      $type: "borderWidth",
      $value: "1px"
    },
    "3": {
      $type: "borderWidth",
      $value: "3px"
    }
  },
  shadow: {
    "100": {
      $type: "boxShadow",
      $value: [
        {
          color: "rgba(0,0,0,0.16)",
          x: "0",
          y: "0",
          blur: "1",
          spread: "0"
        },
        {
          x: "0",
          y: "1",
          blur: "2",
          spread: "0",
          color: "rgba(0,0,0,0.12)"
        }
      ]
    },
    "200": {
      $type: "boxShadow",
      $value: [
        {
          color: "rgba(0,0,0,0.15)",
          x: "0",
          y: "0",
          blur: "1",
          spread: "0"
        },
        {
          color: "rgba(0,0,0,0.12)",
          x: "0",
          y: "1",
          blur: "2",
          spread: "0"
        },
        {
          x: "0",
          y: "2",
          blur: "4",
          spread: "0",
          color: "rgba(0,0,0,0.1)"
        }
      ]
    },
    "300": {
      $type: "boxShadow",
      $value: [
        {
          color: "rgba(0,0,0,0.14)",
          x: "0",
          y: "0",
          blur: "1",
          spread: "0"
        },
        {
          color: "rgba(0,0,0,0.12)",
          x: "0",
          y: "2",
          blur: "4",
          spread: "0"
        },
        {
          x: "0",
          y: "4",
          blur: "8",
          spread: "0",
          color: "rgba(0,0,0,0.12)"
        }
      ]
    },
    "400": {
      $type: "boxShadow",
      $value: [
        {
          color: "rgba(0,0,0,0.13)",
          x: "0",
          y: "0",
          blur: "1",
          spread: "0"
        },
        {
          color: "rgba(0,0,0,0.13)",
          x: "0",
          y: "3",
          blur: "5",
          spread: "0"
        },
        {
          x: "0",
          y: "6",
          blur: "12",
          spread: "0",
          color: "rgba(0,0,0,0.14)"
        }
      ]
    },
    "500": {
      $type: "boxShadow",
      $value: [
        {
          color: "rgba(0,0,0,0.12)",
          x: "0",
          y: "0",
          blur: "1",
          spread: "0"
        },
        {
          color: "rgba(0,0,0,0.16)",
          x: "0",
          y: "4",
          blur: "8",
          spread: "0"
        },
        {
          x: "0",
          y: "12",
          blur: "24",
          spread: "0",
          color: "rgba(0,0,0,0.16)"
        }
      ]
    }
  },
  opacity: {
    "30": {
      $type: "opacity",
      $value: "30%"
    }
  }
};

// src/tokens/template/design-tokens/primitives/modes/size/global.json with { type: 'json' }
var global_default = {
  _size: {
    "0": {
      $type: "dimension",
      $value: "floor({_size.unit} * 0)"
    },
    "1": {
      $type: "dimension",
      $value: "floor({_size.unit} * 1)"
    },
    "2": {
      $type: "dimension",
      $value: "floor({_size.unit} * 2)"
    },
    "3": {
      $type: "dimension",
      $value: "floor({_size.unit} * 3)"
    },
    "4": {
      $type: "dimension",
      $value: "floor({_size.unit} * 4)"
    },
    "5": {
      $type: "dimension",
      $value: "floor({_size.unit} * 5)"
    },
    "6": {
      $type: "dimension",
      $value: "floor({_size.unit} * 6)"
    },
    "7": {
      $type: "dimension",
      $value: "floor({_size.unit} * 7)"
    },
    "8": {
      $type: "dimension",
      $value: "floor({_size.unit} * 8)"
    },
    "9": {
      $type: "dimension",
      $value: "floor({_size.unit} * 9)"
    },
    "10": {
      $type: "dimension",
      $value: "floor({_size.unit} * 10)"
    },
    "11": {
      $type: "dimension",
      $value: "floor({_size.unit} * 11)"
    },
    "12": {
      $type: "dimension",
      $value: "floor({_size.unit} * 12)"
    },
    "13": {
      $type: "dimension",
      $value: "floor({_size.unit} * 13)"
    },
    "14": {
      $type: "dimension",
      $value: "floor({_size.unit} * 14)"
    },
    "15": {
      $type: "dimension",
      $value: "floor({_size.unit} * 15)"
    },
    "18": {
      $type: "dimension",
      $value: "floor({_size.unit} * 18)"
    },
    "22": {
      $type: "dimension",
      $value: "floor({_size.unit} * 22)"
    },
    "26": {
      $type: "dimension",
      $value: "floor({_size.unit} * 26)"
    },
    "30": {
      $type: "dimension",
      $value: "floor({_size.unit} * 30)"
    },
    "mode-font-size": {
      $type: "number",
      $value: "{size._mode-font-size}"
    },
    base: {
      $type: "number",
      $value: "{size._base}"
    },
    step: {
      $type: "number",
      $value: "{size._step}"
    },
    unit: {
      $type: "number",
      $value: "{_size.step} / {_size.base} * {_size.mode-font-size}"
    }
  }
};

// src/tokens/template/design-tokens/primitives/modes/size/large.json with { type: 'json' }
var large_default = {
  size: {
    "_mode-font-size": {
      $type: "dimension",
      $value: "21"
    },
    _base: {
      $type: "dimension",
      $value: "18"
    },
    _step: {
      $type: "dimension",
      $value: "4"
    }
  }
};

// src/tokens/template/design-tokens/primitives/modes/size/medium.json with { type: 'json' }
var medium_default = {
  size: {
    "_mode-font-size": {
      $type: "dimension",
      $value: "18"
    },
    _base: {
      $type: "dimension",
      $value: "18"
    },
    _step: {
      $type: "dimension",
      $value: "4"
    }
  }
};

// src/tokens/template/design-tokens/primitives/modes/size/small.json with { type: 'json' }
var small_default = {
  size: {
    "_mode-font-size": {
      $type: "dimension",
      $value: "16"
    },
    _base: {
      $type: "dimension",
      $value: "18"
    },
    _step: {
      $type: "dimension",
      $value: "4"
    }
  }
};

// src/tokens/template/design-tokens/primitives/modes/typography/size/large.json with { type: 'json' }
var large_default2 = {
  "line-height": {
    sm: {
      $type: "lineHeights",
      $value: "130%"
    },
    md: {
      $type: "lineHeights",
      $value: "150%"
    },
    lg: {
      $type: "lineHeights",
      $value: "170%"
    }
  },
  "font-size": {
    "1": {
      $type: "fontSizes",
      $value: "13"
    },
    "2": {
      $type: "fontSizes",
      $value: "16"
    },
    "3": {
      $type: "fontSizes",
      $value: "18"
    },
    "4": {
      $type: "fontSizes",
      $value: "21"
    },
    "5": {
      $type: "fontSizes",
      $value: "24"
    },
    "6": {
      $type: "fontSizes",
      $value: "30"
    },
    "7": {
      $type: "fontSizes",
      $value: "36"
    },
    "8": {
      $type: "fontSizes",
      $value: "48"
    },
    "9": {
      $type: "fontSizes",
      $value: "60"
    },
    "10": {
      $type: "fontSizes",
      $value: "72"
    }
  },
  "letter-spacing": {
    "1": {
      $type: "letterSpacing",
      $value: "-1%"
    },
    "2": {
      $type: "letterSpacing",
      $value: "-0.5%"
    },
    "3": {
      $type: "letterSpacing",
      $value: "-0.25%"
    },
    "4": {
      $type: "letterSpacing",
      $value: "-0.15%"
    },
    "5": {
      $type: "letterSpacing",
      $value: "0%"
    },
    "6": {
      $type: "letterSpacing",
      $value: "0.15%"
    },
    "7": {
      $type: "letterSpacing",
      $value: "0.25%"
    },
    "8": {
      $type: "letterSpacing",
      $value: "0.5%"
    },
    "9": {
      $type: "letterSpacing",
      $value: "1.5%"
    }
  }
};

// src/tokens/template/design-tokens/primitives/modes/typography/size/medium.json with { type: 'json' }
var medium_default2 = {
  "line-height": {
    sm: {
      $type: "lineHeights",
      $value: "130%"
    },
    md: {
      $type: "lineHeights",
      $value: "150%"
    },
    lg: {
      $type: "lineHeights",
      $value: "170%"
    }
  },
  "font-size": {
    "1": {
      $type: "fontSizes",
      $value: "12"
    },
    "2": {
      $type: "fontSizes",
      $value: "14"
    },
    "3": {
      $type: "fontSizes",
      $value: "16"
    },
    "4": {
      $type: "fontSizes",
      $value: "18"
    },
    "5": {
      $type: "fontSizes",
      $value: "21"
    },
    "6": {
      $type: "fontSizes",
      $value: "24"
    },
    "7": {
      $type: "fontSizes",
      $value: "30"
    },
    "8": {
      $type: "fontSizes",
      $value: "36"
    },
    "9": {
      $type: "fontSizes",
      $value: "48"
    },
    "10": {
      $type: "fontSizes",
      $value: "60"
    }
  },
  "letter-spacing": {
    "1": {
      $type: "letterSpacing",
      $value: "-1%"
    },
    "2": {
      $type: "letterSpacing",
      $value: "-0.5%"
    },
    "3": {
      $type: "letterSpacing",
      $value: "-0.25%"
    },
    "4": {
      $type: "letterSpacing",
      $value: "-0.15%"
    },
    "5": {
      $type: "letterSpacing",
      $value: "0%"
    },
    "6": {
      $type: "letterSpacing",
      $value: "0.15%"
    },
    "7": {
      $type: "letterSpacing",
      $value: "0.25%"
    },
    "8": {
      $type: "letterSpacing",
      $value: "0.5%"
    },
    "9": {
      $type: "letterSpacing",
      $value: "1.5%"
    }
  }
};

// src/tokens/template/design-tokens/primitives/modes/typography/size/small.json with { type: 'json' }
var small_default2 = {
  "line-height": {
    sm: {
      $type: "lineHeights",
      $value: "130%"
    },
    md: {
      $type: "lineHeights",
      $value: "150%"
    },
    lg: {
      $type: "lineHeights",
      $value: "170%"
    }
  },
  "font-size": {
    "1": {
      $type: "fontSizes",
      $value: "11"
    },
    "2": {
      $type: "fontSizes",
      $value: "13"
    },
    "3": {
      $type: "fontSizes",
      $value: "14"
    },
    "4": {
      $type: "fontSizes",
      $value: "16"
    },
    "5": {
      $type: "fontSizes",
      $value: "18"
    },
    "6": {
      $type: "fontSizes",
      $value: "21"
    },
    "7": {
      $type: "fontSizes",
      $value: "24"
    },
    "8": {
      $type: "fontSizes",
      $value: "30"
    },
    "9": {
      $type: "fontSizes",
      $value: "36"
    },
    "10": {
      $type: "fontSizes",
      $value: "48"
    }
  },
  "letter-spacing": {
    "1": {
      $type: "letterSpacing",
      $value: "-1%"
    },
    "2": {
      $type: "letterSpacing",
      $value: "-0.5%"
    },
    "3": {
      $type: "letterSpacing",
      $value: "-0.25%"
    },
    "4": {
      $type: "letterSpacing",
      $value: "-0.15%"
    },
    "5": {
      $type: "letterSpacing",
      $value: "0%"
    },
    "6": {
      $type: "letterSpacing",
      $value: "0.15%"
    },
    "7": {
      $type: "letterSpacing",
      $value: "0.25%"
    },
    "8": {
      $type: "letterSpacing",
      $value: "0.5%"
    },
    "9": {
      $type: "letterSpacing",
      $value: "1.5%"
    }
  }
};

// src/tokens/template/design-tokens/semantic/style.json with { type: 'json' }
var style_default = {
  typography: {
    heading: {
      "2xl": {
        $type: "typography",
        $value: {
          fontFamily: "{font-family}",
          fontWeight: "{font-weight.medium}",
          lineHeight: "{line-height.sm}",
          fontSize: "{font-size.10}",
          letterSpacing: "{letter-spacing.1}"
        }
      },
      xl: {
        $type: "typography",
        $value: {
          fontFamily: "{font-family}",
          fontWeight: "{font-weight.medium}",
          lineHeight: "{line-height.sm}",
          fontSize: "{font-size.9}",
          letterSpacing: "{letter-spacing.1}"
        }
      },
      lg: {
        $type: "typography",
        $value: {
          fontFamily: "{font-family}",
          fontWeight: "{font-weight.medium}",
          lineHeight: "{line-height.sm}",
          fontSize: "{font-size.8}",
          letterSpacing: "{letter-spacing.2}"
        }
      },
      md: {
        $type: "typography",
        $value: {
          fontFamily: "{font-family}",
          fontWeight: "{font-weight.medium}",
          lineHeight: "{line-height.sm}",
          fontSize: "{font-size.7}",
          letterSpacing: "{letter-spacing.3}"
        }
      },
      sm: {
        $type: "typography",
        $value: {
          fontFamily: "{font-family}",
          fontWeight: "{font-weight.medium}",
          lineHeight: "{line-height.sm}",
          fontSize: "{font-size.6}",
          letterSpacing: "{letter-spacing.5}"
        }
      },
      xs: {
        $type: "typography",
        $value: {
          fontFamily: "{font-family}",
          fontWeight: "{font-weight.medium}",
          lineHeight: "{line-height.sm}",
          fontSize: "{font-size.5}",
          letterSpacing: "{letter-spacing.6}"
        }
      },
      "2xs": {
        $type: "typography",
        $value: {
          fontFamily: "{font-family}",
          fontWeight: "{font-weight.medium}",
          lineHeight: "{line-height.sm}",
          fontSize: "{font-size.4}",
          letterSpacing: "{letter-spacing.6}"
        }
      }
    },
    body: {
      xl: {
        $type: "typography",
        $value: {
          fontFamily: "{font-family}",
          fontWeight: "{font-weight.regular}",
          lineHeight: "{line-height.md}",
          fontSize: "{font-size.6}",
          letterSpacing: "{letter-spacing.8}"
        }
      },
      lg: {
        $type: "typography",
        $value: {
          fontFamily: "{font-family}",
          fontWeight: "{font-weight.regular}",
          lineHeight: "{line-height.md}",
          fontSize: "{font-size.5}",
          letterSpacing: "{letter-spacing.8}"
        }
      },
      md: {
        $type: "typography",
        $value: {
          fontFamily: "{font-family}",
          fontWeight: "{font-weight.regular}",
          lineHeight: "{line-height.md}",
          fontSize: "{font-size.4}",
          letterSpacing: "{letter-spacing.8}"
        }
      },
      sm: {
        $type: "typography",
        $value: {
          fontFamily: "{font-family}",
          fontWeight: "{font-weight.regular}",
          lineHeight: "{line-height.md}",
          fontSize: "{font-size.3}",
          letterSpacing: "{letter-spacing.7}"
        }
      },
      xs: {
        $type: "typography",
        $value: {
          fontFamily: "{font-family}",
          fontWeight: "{font-weight.regular}",
          lineHeight: "{line-height.md}",
          fontSize: "{font-size.2}",
          letterSpacing: "{letter-spacing.6}"
        }
      },
      short: {
        xl: {
          $type: "typography",
          $value: {
            fontFamily: "{font-family}",
            fontWeight: "{font-weight.regular}",
            lineHeight: "{line-height.sm}",
            fontSize: "{font-size.6}",
            letterSpacing: "{letter-spacing.8}"
          }
        },
        lg: {
          $type: "typography",
          $value: {
            fontFamily: "{font-family}",
            fontWeight: "{font-weight.regular}",
            lineHeight: "{line-height.sm}",
            fontSize: "{font-size.5}",
            letterSpacing: "{letter-spacing.8}"
          }
        },
        md: {
          $type: "typography",
          $value: {
            fontFamily: "{font-family}",
            fontWeight: "{font-weight.regular}",
            lineHeight: "{line-height.sm}",
            fontSize: "{font-size.4}",
            letterSpacing: "{letter-spacing.8}"
          }
        },
        sm: {
          $type: "typography",
          $value: {
            fontFamily: "{font-family}",
            fontWeight: "{font-weight.regular}",
            lineHeight: "{line-height.sm}",
            fontSize: "{font-size.3}",
            letterSpacing: "{letter-spacing.7}"
          }
        },
        xs: {
          $type: "typography",
          $value: {
            fontFamily: "{font-family}",
            fontWeight: "{font-weight.regular}",
            lineHeight: "{line-height.sm}",
            fontSize: "{font-size.2}",
            letterSpacing: "{letter-spacing.6}"
          }
        }
      },
      long: {
        xl: {
          $type: "typography",
          $value: {
            fontFamily: "{font-family}",
            fontWeight: "{font-weight.regular}",
            lineHeight: "{line-height.lg}",
            fontSize: "{font-size.6}",
            letterSpacing: "{letter-spacing.8}"
          }
        },
        lg: {
          $type: "typography",
          $value: {
            fontFamily: "{font-family}",
            fontWeight: "{font-weight.regular}",
            lineHeight: "{line-height.lg}",
            fontSize: "{font-size.5}",
            letterSpacing: "{letter-spacing.8}"
          }
        },
        md: {
          $type: "typography",
          $value: {
            fontFamily: "{font-family}",
            fontWeight: "{font-weight.regular}",
            lineHeight: "{line-height.lg}",
            fontSize: "{font-size.4}",
            letterSpacing: "{letter-spacing.8}"
          }
        },
        sm: {
          $type: "typography",
          $value: {
            fontFamily: "{font-family}",
            fontWeight: "{font-weight.regular}",
            lineHeight: "{line-height.lg}",
            fontSize: "{font-size.3}",
            letterSpacing: "{letter-spacing.7}"
          }
        },
        xs: {
          $type: "typography",
          $value: {
            fontFamily: "{font-family}",
            fontWeight: "{font-weight.regular}",
            lineHeight: "{line-height.lg}",
            fontSize: "{font-size.2}",
            letterSpacing: "{letter-spacing.6}"
          }
        }
      }
    }
  },
  opacity: {
    disabled: {
      $type: "opacity",
      $value: "{opacity.30}"
    }
  },
  "border-width": {
    default: {
      $type: "borderWidth",
      $value: "{border-width.1}"
    },
    focus: {
      $type: "borderWidth",
      $value: "{border-width.3}"
    }
  },
  shadow: {
    xs: {
      $type: "boxShadow",
      $value: "{shadow.100}"
    },
    sm: {
      $type: "boxShadow",
      $value: "{shadow.200}"
    },
    md: {
      $type: "boxShadow",
      $value: "{shadow.300}"
    },
    lg: {
      $type: "boxShadow",
      $value: "{shadow.400}"
    },
    xl: {
      $type: "boxShadow",
      $value: "{shadow.500}"
    }
  },
  "border-radius": {
    sm: {
      $type: "dimension",
      $value: "{border-radius.1}"
    },
    md: {
      $type: "dimension",
      $value: "{border-radius.2}"
    },
    lg: {
      $type: "dimension",
      $value: "{border-radius.3}"
    },
    xl: {
      $type: "dimension",
      $value: "{border-radius.4}"
    },
    default: {
      $type: "dimension",
      $value: "{border-radius.5}"
    },
    full: {
      $type: "dimension",
      $value: "{border-radius.6}"
    }
  },
  size: {
    "0": {
      $type: "dimension",
      $value: "{_size.0}"
    },
    "1": {
      $type: "dimension",
      $value: "{_size.1}"
    },
    "2": {
      $type: "dimension",
      $value: "{_size.2}"
    },
    "3": {
      $type: "dimension",
      $value: "{_size.3}"
    },
    "4": {
      $type: "dimension",
      $value: "{_size.4}"
    },
    "5": {
      $type: "dimension",
      $value: "{_size.5}"
    },
    "6": {
      $type: "dimension",
      $value: "{_size.6}"
    },
    "7": {
      $type: "dimension",
      $value: "{_size.7}"
    },
    "8": {
      $type: "dimension",
      $value: "{_size.8}"
    },
    "9": {
      $type: "dimension",
      $value: "{_size.9}"
    },
    "10": {
      $type: "dimension",
      $value: "{_size.10}"
    },
    "11": {
      $type: "dimension",
      $value: "{_size.11}"
    },
    "12": {
      $type: "dimension",
      $value: "{_size.12}"
    },
    "13": {
      $type: "dimension",
      $value: "{_size.13}"
    },
    "14": {
      $type: "dimension",
      $value: "{_size.14}"
    },
    "15": {
      $type: "dimension",
      $value: "{_size.15}"
    },
    "18": {
      $type: "dimension",
      $value: "{_size.18}"
    },
    "22": {
      $type: "dimension",
      $value: "{_size.22}"
    },
    "26": {
      $type: "dimension",
      $value: "{_size.26}"
    },
    "30": {
      $type: "dimension",
      $value: "{_size.30}"
    }
  }
};

// src/tokens/create/defaults.ts
var defaultTokens = {
  "primitives/globals": globals_default,
  "primitives/modes/size/small": small_default,
  "primitives/modes/size/medium": medium_default,
  "primitives/modes/size/large": large_default,
  "primitives/modes/size/global": global_default,
  "primitives/modes/typography/size/small": small_default2,
  "primitives/modes/typography/size/medium": medium_default2,
  "primitives/modes/typography/size/large": large_default2,
  "semantic/style": style_default
};
var getDefaultTokens = (tokenSets) => Object.entries(R.pick(tokenSets, defaultTokens));
var getDefaultToken = (tokenPath) => {
  return [tokenPath, defaultTokens[tokenPath]];
};
export {
  getDefaultToken,
  getDefaultTokens
};
