// @digdir/designsystemet/dist/src/tokens/create/generators/semantic.js
// src/tokens/create/generators/semantic.ts
import * as R2 from "ramda";

// src/colors/colorMetadata.ts
import * as R from "ramda";
var baseColors = {
  info: "#0A71C0",
  success: "#068718",
  warning: "#EA9B1B",
  danger: "#C01B1B"
};
var colorMetadata = {
  "background-default": {
    number: 1,
    name: "background-default",
    group: "background",
    displayName: "Background Default",
    description: {
      long: "Background Default er den mest n\xF8ytrale bakgrunnsfargen.",
      short: "Standard bakgrunnsfarge."
    },
    luminance: {
      light: 1,
      dark: 9e-3,
      contrast: 1e-3
    }
  },
  "background-tinted": {
    number: 2,
    name: "background-tinted",
    group: "background",
    displayName: "Background Tinted",
    description: {
      long: "Background Tinted er en bakgrunnsfarge som har et hint av farge i seg.",
      short: "Bakgrunn med et hint av farge i seg."
    },
    luminance: {
      light: 0.9,
      dark: 0.014,
      contrast: 65e-4
    }
  },
  "surface-default": {
    number: 3,
    name: "surface-default",
    group: "surface",
    displayName: "Surface Default",
    description: {
      long: "Surface Default brukes p\xE5 flater som ligger opp\xE5 bakgrunnsfargene. Dette er den mest n\xF8ytrale surface fargen.",
      short: "Standardfarge for overflater / komponenter."
    },
    luminance: {
      light: 1,
      dark: 0.021,
      contrast: 0.015
    }
  },
  "surface-tinted": {
    number: 4,
    name: "surface-tinted",
    group: "surface",
    displayName: "Surface Tinted",
    description: {
      long: "Surface Tinted brukes p\xE5 flater som ligger opp\xE5 bakgrunnsfargene. Denne har et hint av farge i seg.",
      short: "Overflater / komponenter med et hint av farge i seg."
    },
    luminance: {
      light: 0.81,
      dark: 0.027,
      contrast: 0.015
    }
  },
  "surface-hover": {
    number: 5,
    name: "surface-hover",
    group: "surface",
    displayName: "Surface Hover",
    description: {
      long: "Surface Hover brukes p\xE5 interaktive flater som ligger opp\xE5 bakgrunnsfargene i en hover state.",
      short: "Hover-farge til overflater / komponenter."
    },
    luminance: {
      light: 0.7,
      dark: 0.036,
      contrast: 0.028
    }
  },
  "surface-active": {
    number: 6,
    name: "surface-active",
    group: "surface",
    displayName: "Surface Active",
    description: {
      long: "Surface Active brukes p\xE5 interaktive flater som ligger opp\xE5 bakgrunnsfargene i en active state.",
      short: "Active-farge til overflater / komponenter."
    },
    luminance: {
      light: 0.59,
      dark: 0.056,
      contrast: 0.045
    }
  },
  "border-subtle": {
    number: 7,
    name: "border-subtle",
    group: "border",
    displayName: "Border Subtle",
    description: {
      long: "Border Subtle er den lyseste border-fargen og brukes for \xE5 skille elementer fra hverandre.",
      short: "Border-farge med lav kontrast til dekorativ bruk (skillelinjer)."
    },
    luminance: {
      light: 0.5,
      dark: 0.08,
      contrast: 0.26
    }
  },
  "border-default": {
    number: 8,
    name: "border-default",
    group: "border",
    displayName: "Border Default",
    description: {
      long: "Border Default er en border-farge som brukes n\xE5r man \xF8nsker god kontrast mot bakgrunnsfargene.",
      short: "Standard border-farge til skjemakomponenter og meningsb\xE6rende elementer."
    },
    luminance: {
      light: 0.19,
      dark: 0.22,
      contrast: 0.4
    }
  },
  "border-strong": {
    number: 9,
    name: "border-strong",
    group: "border",
    displayName: "Border Strong",
    description: {
      long: "Border Strong er den m\xF8rkeste border-fargen og brukes n\xE5r man \xF8nsker en veldig tydelig og sterk border.",
      short: "Border-farge med h\xF8y kontrast for ekstra synlighet."
    },
    luminance: {
      light: 0.11,
      dark: 0.39,
      contrast: 0.6
    }
  },
  "text-subtle": {
    number: 10,
    name: "text-subtle",
    group: "text",
    displayName: "Text Subtle",
    description: {
      long: "Text Subtle er den lyseste tekstfargen og brukes p\xE5 tekst som skal v\xE6re litt mindre synlig eller for \xE5 skape variasjon i typografien.",
      short: "Tekst- og ikonfarge med lavere kontrast."
    },
    luminance: {
      light: 0.11,
      dark: 0.39,
      contrast: 0.57
    }
  },
  "text-default": {
    number: 11,
    name: "text-default",
    group: "text",
    displayName: "Text Default",
    description: {
      long: "Text Default er den m\xF8rkeste tekstfargen og brukes p\xE5 tekst som skal v\xE6re mest synlig. Denne fargen b\xF8r brukes p\xE5 mesteparten av teksten p\xE5 en side.",
      short: "Tekst- og ikonfarge med h\xF8y kontrast og god synlighet."
    },
    luminance: {
      light: 0.0245,
      dark: 0.84,
      contrast: 0.86
    }
  },
  "base-default": {
    number: 12,
    name: "base-default",
    group: "base",
    displayName: "Base Default",
    description: {
      long: "Base Default fargen f\xE5r den samme hex koden som fargen som er valgt i verkt\xF8yet. Brukes ofte som farge p\xE5 viktige elementer og p\xE5 flater som skal fange brukerens oppmerksomhet.",
      short: "Standardfarge for solide bakgrunner."
    },
    luminance: {
      light: 1,
      dark: 1,
      contrast: 1
    }
  },
  "base-hover": {
    number: 13,
    name: "base-hover",
    group: "base",
    displayName: "Base Hover",
    description: {
      long: "Base Hover brukes som hover farge p\xE5 elementer som bruker Base Default fargen.",
      short: "Hover-farge for solide bakgrunner."
    },
    luminance: {
      light: 1,
      dark: 1,
      contrast: 1
    }
  },
  "base-active": {
    number: 14,
    name: "base-active",
    group: "base",
    displayName: "Base Active",
    description: {
      long: "Base Active brukes som active farge p\xE5 elementer som bruker Base Default fargen.",
      short: "Active-farge for solide bakgrunner."
    },
    luminance: {
      light: 1,
      dark: 1,
      contrast: 1
    }
  },
  "base-contrast-subtle": {
    number: 15,
    name: "base-contrast-subtle",
    group: "base",
    displayName: "Contrast Subtle",
    description: {
      long: "Contrast Subtle brukes som en viktig meningsb\xE6rende farge opp\xE5 Base Default fargen.",
      short: "Farge med god kontrast mot Base-default."
    },
    luminance: {
      light: 1,
      dark: 1,
      contrast: 1
    }
  },
  "base-contrast-default": {
    number: 16,
    name: "base-contrast-default",
    group: "base",
    displayName: "Contrast Default",
    description: {
      long: "Contrast Default brukes som en viktig meningsb\xE6rende farge opp\xE5 alle Base fargane.",
      short: "Farge med god kontrast mot Base-default og Base-hover."
    },
    luminance: {
      light: 1,
      dark: 1,
      contrast: 1
    }
  }
};
var colorMetadataByNumber = R.indexBy((metadata) => metadata.number, Object.values(colorMetadata));
var colorNames = Object.keys(colorMetadata);
var baseColorNames = Object.keys(baseColors);

// src/tokens/template/design-tokens/semantic/color.base.template.json with { type: 'json' }
var color_base_template_default = {
  color: {
    focus: {
      inner: {
        $type: "color",
        $value: "{color.neutral.background-default}"
      },
      outer: {
        $type: "color",
        $value: "{color.neutral.text-default}"
      }
    }
  },
  link: {
    color: {
      visited: {
        $type: "color",
        $value: "{color.link.visited}"
      }
    }
  }
};

// src/tokens/template/design-tokens/semantic/color.template.json with { type: 'json' }
var color_template_default = {
  "background-default": {
    $type: "color",
    $value: "{color.<color>.1}"
  },
  "background-tinted": {
    $type: "color",
    $value: "{color.<color>.2}"
  },
  "surface-default": {
    $type: "color",
    $value: "{color.<color>.3}"
  },
  "surface-tinted": {
    $type: "color",
    $value: "{color.<color>.4}"
  },
  "surface-hover": {
    $type: "color",
    $value: "{color.<color>.5}"
  },
  "surface-active": {
    $type: "color",
    $value: "{color.<color>.6}"
  },
  "border-subtle": {
    $type: "color",
    $value: "{color.<color>.7}"
  },
  "border-default": {
    $type: "color",
    $value: "{color.<color>.8}"
  },
  "border-strong": {
    $type: "color",
    $value: "{color.<color>.9}"
  },
  "text-subtle": {
    $type: "color",
    $value: "{color.<color>.10}"
  },
  "text-default": {
    $type: "color",
    $value: "{color.<color>.11}"
  },
  "base-default": {
    $type: "color",
    $value: "{color.<color>.12}"
  },
  "base-hover": {
    $type: "color",
    $value: "{color.<color>.13}"
  },
  "base-active": {
    $type: "color",
    $value: "{color.<color>.14}"
  },
  "base-contrast-subtle": {
    $type: "color",
    $value: "{color.<color>.15}"
  },
  "base-contrast-default": {
    $type: "color",
    $value: "{color.<color>.16}"
  }
};

// src/tokens/template/design-tokens/semantic/modes/color.template.json with { type: 'json' }
var color_template_default2 = {
  "background-default": {
    $type: "color",
    $value: "{color.<color>.background-default}"
  },
  "background-tinted": {
    $type: "color",
    $value: "{color.<color>.background-tinted}"
  },
  "surface-default": {
    $type: "color",
    $value: "{color.<color>.surface-default}"
  },
  "surface-tinted": {
    $type: "color",
    $value: "{color.<color>.surface-tinted}"
  },
  "surface-hover": {
    $type: "color",
    $value: "{color.<color>.surface-hover}"
  },
  "surface-active": {
    $type: "color",
    $value: "{color.<color>.surface-active}"
  },
  "border-subtle": {
    $type: "color",
    $value: "{color.<color>.border-subtle}"
  },
  "border-default": {
    $type: "color",
    $value: "{color.<color>.border-default}"
  },
  "border-strong": {
    $type: "color",
    $value: "{color.<color>.border-strong}"
  },
  "text-subtle": {
    $type: "color",
    $value: "{color.<color>.text-subtle}"
  },
  "text-default": {
    $type: "color",
    $value: "{color.<color>.text-default}"
  },
  "base-default": {
    $type: "color",
    $value: "{color.<color>.base-default}"
  },
  "base-hover": {
    $type: "color",
    $value: "{color.<color>.base-hover}"
  },
  "base-active": {
    $type: "color",
    $value: "{color.<color>.base-active}"
  },
  "base-contrast-subtle": {
    $type: "color",
    $value: "{color.<color>.base-contrast-subtle}"
  },
  "base-contrast-default": {
    $type: "color",
    $value: "{color.<color>.base-contrast-default}"
  }
};

// src/tokens/create/generators/semantic.ts
var generateSemantic = (colors, _themeName) => {
  const mainColorNames = Object.keys(colors.main);
  const supportColorNames = Object.keys(colors.support);
  const modes = {
    "main-color": {},
    "support-color": {}
  };
  const categories = [
    ["main-color", mainColorNames],
    ["support-color", supportColorNames]
  ];
  for (const [colorCategory, colorNames2] of categories) {
    for (const colorName of colorNames2) {
      const category = colorCategory.replace("-color", "");
      const customColorTokens = {
        color: {
          [category]: JSON.parse(
            JSON.stringify(
              color_template_default2,
              (key, value) => {
                if (key === "$value") {
                  return value.replace("<color>", colorName);
                }
                return value;
              },
              2
            )
          )
        }
      };
      modes[colorCategory][colorName] = customColorTokens;
    }
  }
  const customColors = [...mainColorNames, "neutral", ...supportColorNames];
  const allColors = [...customColors, ...baseColorNames];
  const semanticColorTokens = allColors.map(
    (colorName) => [
      colorName,
      R2.map((x) => ({ ...x, $value: x.$value.replace("<color>", colorName) }), color_template_default)
    ]
  );
  const color = {
    ...color_base_template_default,
    color: {
      ...Object.fromEntries(semanticColorTokens),
      ...color_base_template_default.color
    }
  };
  return {
    modes,
    color
  };
};
export {
  generateSemantic
};
